e2etest/configurationstatetests/configurationstate.go

```diff
@@ -0,0 +1,249 @@
+		Result:       metallbv1beta1.ConfigurationResultValid,
+		ErrorSummary: "",
+	}
+)
+
+var _ = ginkgo.Describe("ConfigurationState", func() {
+	var allNodes *corev1.NodeList
+
+	ginkgo.BeforeEach(func() {
+		ginkgo.By("Clearing any previous configuration")
+		err := ConfigUpdater.Clean()
+		Expect(err).NotTo(HaveOccurred())
+
+		ginkgo.By("Getting all nodes")
+		cs := k8sclient.New()
+		allNodes, err = cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(allNodes.Items)).To(BeNumerically(">", 0))
+
+		ginkgo.By("Verifying all ConfigurationStates exist and are valid")
+		Eventually(func() error {
+			return allStatesExist(allNodes)
+		}, 30*time.Second, 2*time.Second).Should(Succeed())
+	})
+
+	ginkgo.AfterEach(func() {
+		if ginkgo.CurrentSpecReport().Failed() {
+			k8s.DumpInfo(Reporter, ginkgo.CurrentSpecReport().LeafNodeText)
+		}
+	})
+
+	ginkgo.It("speaker should have invalid result when BGPPeer references secret with wrong type", func() {
+		stateName := "speaker-" + allNodes.Items[0].Name
+		wantStatus := metallbv1beta1.ConfigurationStateStatus{
+			Result:       metallbv1beta1.ConfigurationResultInvalid,
+			ErrorSummary: "configuration error: parsing peer peer1 secret type mismatch on \"metallb-system\"/\"bgp-password\", type \"kubernetes.io/basic-auth\" is expected \nfailed to parse peer peer1 password secret",
+		}
+
+		ginkgo.By("Creating secret with wrong type and BGPPeer referencing it")
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "bgp-password",
+				Namespace: metallb.Namespace,
+			},
+			Type: corev1.SecretTypeOpaque,
+			StringData: map[string]string{
+				"password": "mypassword",
+			},
+		}
+		err := ConfigUpdater.Client().Create(context.Background(), secret)
+		Expect(err).NotTo(HaveOccurred())
+		ginkgo.DeferCleanup(func() {
+			ConfigUpdater.Client().Delete(context.Background(), secret)
+		})
+
+		resources := config.Resources{
+			Peers: []metallbv1beta2.BGPPeer{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "peer1",
+					},
+					Spec: metallbv1beta2.BGPPeerSpec{
+						MyASN:   64512,
+						ASN:     64513,
+						Address: "192.168.1.1",
+						PasswordSecret: corev1.SecretReference{
+							Name: "bgp-password",
+						},
+					},
+				},
+			},
+		}
+
+		err = ConfigUpdater.Update(resources)
+		Expect(err).NotTo(HaveOccurred())
+
+		ginkgo.By("Verifying status has invalid result with error message")
+		Eventually(func() error {
+			return stateMatches(stateName, wantStatus)
+		}, 30*time.Second, 5*time.Second).Should(Succeed())
+
+		ginkgo.By("Recreating secret with correct type")
+		err = ConfigUpdater.Client().Delete(context.Background(), secret) // field is immutable, need to delete first
+		Expect(err).NotTo(HaveOccurred())
+		secret.ResourceVersion = ""
+		secret.UID = ""
+		secret.Type = corev1.SecretTypeBasicAuth
+		err = ConfigUpdater.Client().Create(context.Background(), secret)
```

Comment:
I think if we don't ensure the secret is deleted, this might flake failing because teh secret is still being deleted